_posts/2023-01-03-encoders-and-multiplexers.md

```diff
@@ -0,0 +1,93 @@
+
+![01](/assets/images/encoders-multiplexers/01.png)
+Fonte: Autor. Utilizado o IBM Quantum Composer.
+
+![10](/assets/images/encoders-multiplexers/10.png)
+Fonte: Autor. Utilizado o IBM Quantum Composer.
+
+![11](/assets/images/encoders-multiplexers/11.png)
+Fonte: Autor. Utilizado o IBM Quantum Composer.
+
+Onde o último qubit (ou primeiro qubit no Composer) é a resposta da operação. Uma implementação em Python usando o Qiskit está presente em [ref.I].
+
+Agora poderiammos montar o codificador binário 8:3 se o IBM Quantum Composer nos fornecesse uma ferramenta direta para trabalharmos com mais de 7 qubtis. Como não é o caso, teremos que nos limitar, por enquanto, a esquemas e linhas de código. Uma implementação em Python utilizando o Qiskit e o IBM podem ser encontradas em [ref.I]. A construção do codificador binário 8:3 fica, conforme [ref.I]:
+
+![encoder](/assets/images/encoders-multiplexers/encoderfinal.png)
+Fonte: Narula, Hridey & Behera, Bikash & Panigrahi, Prasanta. (2020). Implementing Quantum Data Processing Circuits Using IBM Quantum Experience Platform. 10.13140/RG.2.2.36062.38729
+
+Onde as linhas q16 correspondem aos inputs, q17 os outputs, q18 e q19 à qubits auxiliares usados como buffer e validação. Foram utilizados 16 qubits porque queriam preservar os valores de entrada.
+
+### Multiplexador 4:1
+
+Para implementar um multiplexer precisaremos das portas quânticas que reproduzem o efeito das portas clássicas OR e AND. Entretanto, a porta OR já foi estudada na seção anterior e a porta AND é equivalente, em questão de tabela verdade, a porta Toffoli desde que o target qubit seja 0. De fato, fixando o terceiro qubit de entrada como zero no Qiskit:
+
+![000](/assets/images/encoders-multiplexers/000.png)
+Fonte: Autor. Utilizado o IBM Quantum Composer.
+
+![010](/assets/images/encoders-multiplexers/010.png)
+Fonte: Autor. Utilizado o IBM Quantum Composer.
+
+![100](/assets/images/encoders-multiplexers/100.png)
+Fonte: Autor. Utilizado o IBM Quantum Composer.
+
+![110](/assets/images/encoders-multiplexers/110.png)
+Fonte: Autor. Utilizado o IBM Quantum Composer.
+
+Para fixar o terceiro qubit como 0 podemos usar uma estrutura condicional que flipe o qubit caso este seja 1. Dessa forma nosso "AND quântico" fica:
+
+![Gate-AND](/assets/images/encoders-multiplexers/quantum_and.png)
+Fonte: Autor. Utilizado o IBM Quantum Composer.
+
+Da mesma forma que para o codificador, não podemos implementar um multiplexador no IBM Quantum Composer. O circuito, caso tivessemos 16 qubits, ficaria:
+![multiplexer](/assets/images/encoders-multiplexers/mux.png)
+Fonte: Narula, Hridey & Behera, Bikash & Panigrahi, Prasanta. (2020). Implementing Quantum Data Processing Circuits Using IBM Quantum Experience Platform. 10.13140/RG.2.2.36062.38729
+
+Note que uma vez criado o codificador e o multiplexador, basta inverter as portas para obtermos o decodificador e o demultiplexador. Estes são detalhados em [ref.I].
```

Comment:
Essa frase me passa a sensação de ter que inverter o circuito, mas na verdade temos que inverter o input (aplicar um not nele).

Reply:
Eu tinha escrito assim com o intuito de inverter os gates/circuito mesmo. Não sei se entendi, apenas aplicando um not no input (os três bits) teríamos a ação do decodificador?

Reply:
Vou refazer a parte de decodificadores e demultiplexadores. Percebi que é mais complicado do que eu pensava. Obrigado pela ajuda!